src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs

```diff
@@ -206,9 +207,36 @@ public override void OnStopActivity(Activity activity, object payload)
 
                 if (!string.IsNullOrEmpty(template))
                 {
-                    // Override the name that was previously set to the path part of URL.
-                    activityToEnrich.DisplayName = template;
-                    activityToEnrich.SetTag(SemanticConventions.AttributeHttpRoute, template);
+                    bool isRootPath = request.Path == "/";
+                    bool isIdOptional = false;
+
+                    // For urls without ID, for e.g. GET all resources of a kind (/api/Values/), we don't need to use template information
+                    if (!isRootPath && request.Path.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries).Length <
+                        template.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries).Length)
+                    {
+                        isIdOptional = true;
+                        activityToEnrich.DisplayName = activity.DisplayName.Substring(1); // Remove '/' from beginning of the activity name for consistency
+                    }
+
+                    // Override the name that was previously set to the path part of URL. If it is a root path or if Id is optional, activity DisplayName
+                    // is already set correctly in OnStartActivity method
+                    if (!isRootPath && !isIdOptional)
+                    {
+                        if (routeData.Values.TryGetValue("controller", out object controller))
+                        {
+                            activityToEnrich.DisplayName = template.Replace("{controller}", controller.ToString());
```

Comment:
I think we could improve this string manipulation to be a bit more performant but we can do that in a follow-up.